test/src/graph/addChildTest.cpp

```diff
@@ -0,0 +1,152 @@
+#include <random>
+#include <gtest/gtest.h>
+#include <utils.hpp>
+#include <RGLExceptions.hpp>
+#include <RGLFields.hpp>
+
+
+using namespace ::testing;
+
+/**
+ * This test is supposed to cover different situations when rgl_graph_add_child is called:
+ * - Parent is a correct / incorrect input to the child (onInputValidate() passes or fails)
+ * - Parent / child differ of being hot / cold (having GraphContext assigned) (onGraphChange() passes or fails)
+ */
+struct GraphAddChild : public RGLAutoCleanupTestWithParam<bool>
+{
+protected:
+	GraphAddChild()
+	{
+		// Prepare a set of test nodes to cover all cases. In some tests, some nodes will remain unused.
+		EXPECT_RGL_SUCCESS(rgl_node_points_from_array(&fromArrayA, points, 1, fields, 1));
+		EXPECT_RGL_SUCCESS(rgl_node_points_from_array(&fromArrayB, points, 1, fields, 1));
+		EXPECT_RGL_SUCCESS(rgl_node_points_yield(&yieldA, fields, 1));
+		EXPECT_RGL_SUCCESS(rgl_node_points_yield(&yieldB, fields, 1));
+	}
+
+	void addChild(rgl_node_t parent, rgl_node_t childYield, bool expectAddChildSuccess = true)
+	{
+		bool correctInput = GetParam();
+		if (!correctInput) {
+			// Modify child node to make it report invalid (incompatible) input
+			EXPECT_RGL_SUCCESS(rgl_node_points_yield(&childYield, wrongFields, 1));
+		}
+
+		if (!correctInput) {
+			EXPECT_THROW(rgl_graph_node_add_child(parent, childYield), InvalidPipeline);
+			SUCCEED(); // Do not proceed when onInputChangeImpl() was supposed to fail
+		} else {
```

Comment:
I get the logic, however it will be super unclear to read this nested test when something will fail.